fix!: update code and tests for pyjwt>=2.0.0

karls opened the pull request:
# Fixes #556 

This PR upgrades pyjwt to 2.0.1, which is a major version upgrade and a breaking change. Namely, pyjwt 2.x drops support for Python 2.x and Python 3.0-3.5. It also changes some user-facing APIs that require changes to twilio-python.

I've made a start on updating the code and fixing broken tests. It would be great to get feedback on the changes in `twilio/jwt/__init__.py:143-144`. Since the `algorithms` kwarg is now a required argument for `jwt.decode(...)` and this library uses both HS256 and RS256, the approach I've taken is to try and get the appropriate algorithm from the particular token's header and fall back to HS256 if it's missing. The alternative approach would be to explicitly pass in `algorithms=["HS256", "RS256"]` without trying to get the algorithm from the header.

That's as far as the "easy" changes go. The harder decision is dropping support for older Python versions and what kind of changes that requires. Happy to help, but I'd need guidance.

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guidelines](https://github.com/twilio/twilio-python/blob/main/CONTRIBUTING.md) and my PR follows them
- [x] I have titled the PR appropriately
- [x] I have updated my branch with the main branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation about the functionality in the appropriate .md file
- [ ] I have added inline documentation to the code I modified

wavenator commented on requirements.txt:
```diff
@@ -3,5 +3,5 @@ mock
-PyJWT==1.7.1
-twine
\ No newline at end of file
+PyJWT>=2.0.0
```
I'd recommend putting a high bar on this dependency. There might be a point in the future where it will become 3.x and will break API.

karls replied:
Good spot, updated. Thanks!

wavenator commented:
If you want to merge this PR you will have to remove Python 3.5 < support. PyJWT does not support Python versions below 3.6.
You will have to edit the `Readme.md` and `.travis.yml`. Try finding more occurrences of lower Python versions.

karls commented:
Hey folks. I've gone through and removed references to Python 2.7, updated the build matrix, updated callsites using `six` for backwards compat and removed some compat code. It would be great to get some extra eyes on this and feedback — I'm not an expert in removing support for old Python versions. :-)

wavenator commented:
Maybe removing the Python version badges from the `Readme.md`?

karls commented:
> Maybe removing the Python version badges from the `Readme.md`?

I think that's generated by shields.io by looking at the supported Python versions as given in PyPI. Once this PR is merged and a new version is pushed to PyPI that supports Python 3.6+, I think that badge will auto-update itself.

Or did you mean just remove it completely?

wavenator commented:
You got it right :)

thinkingserious commented:
Thank you @karls for the PR! And to @wavenator and @amloren1 for your feedback!

Currently, we are still supporting 3.5 and do not support 3.9. That said, both of those items are on our backlog. However, we do not have those items committed to a specific timeframe just yet. I've added a note internally to make sure we revisit this PR once we commit to dropping support for 3.5. 

That said, additional comments and +1s on this PR will help move up the priority.

With best regards,

Elmer

karls commented:
Likewise, thank you for the update @thinkingserious and for feedback @wavenator and @amloren1.

Quite a few people seem to be blocked on this update — it would be great to get at least a rough idea, even a ballpark, of how long the wait for the official release would be.

FWIW, I've started using my fork in production (without issues so far), pinned to [f155dbe](https://github.com/karls/twilio-python/commit/f155dbe82c08eee131722a857a66397a6b01a2b3). I only use a very small subset of the surface area of the library, though — just the AccessToken API and programmable video (rooms, media, etc). I think if there are other brave people willing to test this in production and provide feedback, it'll help with getting this PR and the new release in proper shape in time for the official release.

What do you think @thinkingserious?

thinkingserious commented:
Hello @karls, @amloren1,

Could I please have e more context as to why this issue is a blocker with respect to the `pyjwt` upgrade? I'm trying to make sure I understand the consequence of dropping support of version `1.7.1`, besides the necessity to continue supporting EOL versions of Python. Thank you!

Unfortunately, I don't have a ballpark to share just yet on the official deprecation timeline.

Internal tracking ID: DI-1105

karls commented:
People who need to or want to upgrade to `PyJWT >= 2.0` are unable to do so because `twilio-python` versions 6.51.0 and newer pin `PyJWT` to 1.7.1, which the dependency resolver is unable to reconcile. And `twilio-python` versions earlier than 6.51.0 do not specify `PyJWT` version, which means that you *can* install `PyJWT >= 2.0`, but `twilio-python` will break if your codebase uses AccessTokens (or other classes from `twilio.jwt.*` that depend on encoding and decoding JWTs).

Does that help @thinkingserious?

eshanholtz commented:
@karls Thank you for the additional context and we appreciate the work you're doing to create a fork that unblocks folks! It looks like we can add support for 3.9 now without any issues, so I went ahead and approved and merged @tim-schilling's #563 PR. 

With regards to the `PyJWT >= 2.0` upgrade, since removing the hard version requirement breaks the existing AccessToken functionality, and upgrading the version would require us to drop currently supported Python versions, we're looking at a 1 month+ minimum timeline for rolling to a new major version of `twilio-python`, which would include these changes. This PR and the concerns raised in this thread have definitely accelerated our timeline for this, so please be patient and rest assured that we hear you and are working on this :)

For those of you in the community who are just finding this thread, please note that we use +1's on the [original PR description](https://github.com/twilio/twilio-python/pull/560/#issue-567867779) to help us gauge community demand. Thank you all for your feedback and contributions!

karls commented:
Thank you for the update @eshanholtz. And I'm glad this issue is making its way higher up in the priority list. :-)

I'm going to deploy our app using https://github.com/karls/twilio-python/commit/872f52e74ba26a98ec81722653933724912ea758 today in order to keep making sure things are working. We're on Python 3.8.8 FWIW, no issues so far at least. It would be awesome if other folks would be willing to temporarily test my fork in staging or even in production.

christianbundy commented:
> That said, additional comments and +1s on this PR will help move up the priority.

+1, we're also blocked from upgrading twilio-python because we use pyjwt 2.x.x. Thanks for moving this forward!

hugorodgerbrown commented:
Thank you @karls for this PR - hopefully it gets through the approval process soon enough.

hugorodgerbrown commented:
> It would be awesome if other folks would be willing to temporarily test my fork in staging or even in production.

@karls we have forked your fork (we only permit pinning to our own repos), and have installed the latest commit on that (https://github.com/yunojuno/twilio-python/commit/4889644bf5c36625a7d946700ee97e5ff613d4f7). We've unpinned our other packages that depend on PyJWT (`django-request-token` and `social-auth-core`) so they can upgrade. I'll report back on impact.

NB we only use this lib for SMS (i.e. no JWT) so not a full stress test of the changes.

bharling commented:
+1 For this please - just updated social-auth-core (much needed) and found twilio now broken. This is a major blocker for us right now! thanks in advance :)

hugorodgerbrown commented:
Update on our usage - works fine for us in our staging environment (Heroku). SMS sending only.

hugorodgerbrown commented:
Giving this a bump just in case anyone from Twilio is watching. We're running @karls's fork in production now, and it's working.

bharling commented:
+1 Also running the fork in production and it's working well, would really like to see this in mainline soon if we can!

andersk commented on twilio/jwt/__init__.py:
```diff
@@ -140,7 +134,8 @@ def from_jwt(cls, jwt, key=''):
         try:
-            payload = jwt_lib.decode(bytes(jwt), key, options={
+            alg = jwt_lib.get_unverified_header(jwt).get("alg", "HS256")
+            payload = jwt_lib.decode(jwt, key, algorithms=[alg], options={
```
This is an unsafe pattern. (Just as unsafe as the code you’re replacing, but still unsafe.) The problem is that if you’re expecting an RS256 token, an attacker should not be able to forge the token without access to the signer’s RSA **private** key; however, if you trust the algorithm specified in the token’s unverified header, then the attacker can substitute an HS256 token hashed with the signer’s RSA **public** key. See:

https://www.rfc-editor.org/rfc/rfc8725.html#section-2.1
https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/

Since Twilio [documents](https://www.twilio.com/docs/iam/access-tokens#jwt-format) that the value of `alg` **must** be `"HS256"`, we should just specify `algorithms=["HS256"]` and not allow the token’s unverified header to modify this expectation.

karls replied:
That's a great find — thank you.

Twilio library also uses RS256 in `twilio.jwt.validation.ClientValidationJwt`, which inherits from `twilio.jwt.Jwt`, but upon reading the docs and looking through the code it looks like that class is used to sign HTTP requests to Twilio servers. You upload your public key to Twilio, then use `twilio.http.ValidationClient` to sign requests, as per [the docs](https://www.twilio.com/docs/iam/pkcv/quickstart).

I'm going to change `algorithms` to `["HS256"]` shortly.

karls replied:
Looking through the code again, I don't understand why the choice of algorithm is open to overriding if the docs clearly state that it _has_ to be HS256 or RS256 in `t.j.Jwt` or `t.j.v.ClientValidationJwt`, respectively.

Maybe someone from Twilio can chime in? If the choice of algorithm in those cases is an invariant, we should probably make `algorithm` a class variable and not let the user override it?

thinkingserious replied:
Thanks for the code review @andersk!!

@karls, I'm OK with not letting the user override the choice of algorithm.

karls replied:
I've updated the code. It would be great if @thinkingserious or someone else from Twilio could review the changes.

I had some trouble coming up with a way to completely "hide" the choice of algorithm without rewriting a bunch of code. The constructor for `twilio.jwt.Jwt` currently still allows passing in a custom algorithm, but it defaults to `HS256` as before.

`AccessToken`, `ClientCapabilityToken` and `TaskRouterCapabilityToken` are defined to use `HS256` and `ClientValidationJwt` is defined to use `RS256`.

Library users will now be unable to override the algorithm in `Jwt.to_jwt()`. I have also added an additional check in `Jwt.from_jwt()` to make sure that the algorithm in the token header matches the algorithm designated for the particular token. Technically, you can trigger this if you manually construct a `Jwt` (e.g not via one of the Token classes named above, or `ClientValidationJwt`) with, for instance, `HS512` and then decode it with `Jwt.from_jwt()`. I'm not sure how much of an issue this is in practise.

The assumption I've made is the following — a particular type of token should be decoded using the classmethod of its class. E.g if you create a `ClientCapabilityToken`, you should call `ClientCapabilityToken.from_jwt()` to decode it, not `Jwt.from_jwt()`.

karls commented:
A quick heads-up to people using my fork, I've updated the code as per @andersk's comments and I've added an explanation of the changes here: https://github.com/twilio/twilio-python/pull/560#discussion_r615251847.

We're also running my fork with the latest code (https://github.com/karls/twilio-python/commit/2a323793568d322ad7fc0cb5793ebd05931e8d4a) in production without issues — using only `AccessToken` and programmable video, though.

thinkingserious commented:
Internal tracking ID: DI-1327

karls commented:
Any updates on timelines/progress internally?

shwetha-manvinkurke commented:
@karls it has moved up the backlog and we will update you on the timeline when we have that defined.

JenniferMah reviewed (approved):
Hi @karls thanks for your patience! This feature continues to move up in our backlog as our team prepares to drop support for python 2.7, 3.4, and 3.5. We should have an update on the estimated timeline within the next week.

JenniferMah commented:
Hi Folks!
Thank you for your patience on this. On or after September 22 2021, Twilio will release version 7.0.0 of the Twilio Python helper library. This version will officially drop support for Python versions 2.7, 3.4, and 3.5. Twilio Python version 7.0.0 will also include the upgrade to PyJWT 2.0.
